cs4660/graph/graph.py

```diff
@@ -0,0 +1,240 @@
+    first_line = f.readline()
+    for i in range(int(first_line)):
+        graph.add_node(Node(i))
+    text = f.read()
+    lines = text.split('\n')
+    for line in lines:
+        if len(line) > 0:
+            values = line.split(':')
+            edge = Edge( Node( int( values[0] ) ), Node( int( values[1] ) ), int( values[2] ) )
+            graph.add_edge(edge)
+    return graph
+
+class Node(object):
+    """Node represents basic unit of graph"""
+    def __init__(self, data):
+        self.data = data
+
+    def __str__(self):
+        return 'Node({})'.format(self.data)
+    def __repr__(self):
+        return 'Node({})'.format(self.data)
+
+    def __eq__(self, other_node):
+        return self.data == other_node.data
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash(self.data)
+
+class Edge(object):
+    """Edge represents basic unit of graph connecting between two edges"""
+    def __init__(self, from_node, to_node, weight):
+        self.from_node = from_node
+        self.to_node = to_node
+        self.weight = weight
+    def __str__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+    def __repr__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+
+    def __eq__(self, other_node):
+        return self.from_node == other_node.from_node and self.to_node == other_node.to_node and self.weight == other_node.weight
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.from_node, self.to_node, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        # adjacencyList should be a dictonary of node to edges
+        self.adjacency_list = {}
+
+    def adjacent(self, node_1, node_2):
+        return node_1 in self.adjacency_list and node_2 in self.adjacency_list[node_1]
+
+    def neighbors(self, node):
+        if(node in self.adjacency_list):
+            return self.adjacency_list[node]
+        return []
+
+    def add_node(self, node):
+        if(node in self.adjacency_list):
+            return False
+        self.adjacency_list[node]=[]
+        return True
+
+    def remove_node(self, node):
+        if(node not in self.adjacency_list):
+            return False
+        del self.adjacency_list[node]
+        for each_key in self.adjacency_list:
+            if(node in self.adjacency_list[each_key]):
+                self.adjacency_list[each_key].remove(node)
+        return True
+
+    def add_edge(self, edge):
+        self.add_node(edge.from_node)
+        if(edge.to_node in self.adjacency_list[edge.from_node]):
+            return False
+        self.adjacency_list[edge.from_node].append(edge.to_node)
```

Comment:
This implementation ignores the edge weight. In other word, the implementation is not completed and will fail at upcoming homework 2 in Dijkstra search and A* search.